src/TSL/HOA/Python.hs

```diff
@@ -5,14 +5,51 @@ module TSL.HOA.Python
 where
 
 import Data.List (intercalate)
+import qualified Data.Set as Set
 import qualified Hanoi as H
+import TSL.HOA.Codegen (codegen, splitInputsCellsOutputs)
 import TSL.HOA.Imp
   ( ImpConfig (..),
-    withConfig,
+    cellOutputNextPrefix,
+    withConfig',
   )
 
 implement :: Bool -> H.HOA -> String
-implement = withConfig config
+implement isCounterStrat hoa =
+  let prog = codegen hoa
+      controller = withConfig' config isCounterStrat prog
+      (is', cs', os') = splitInputsCellsOutputs prog
+      is = Set.toList is'
+      cs = Set.toList cs'
+      os = Set.toList os'
+   in "def controller(_inputs_and_cells):\n"
+        -- inputs and cells
+        ++ ( if not (null (is ++ cs))
+               then
+                 indent 1
+                   ++ commas (is ++ cs)
+                   ++ " = itemgetter("
+                   ++ commas (map wrapQuotes (is ++ cs))
+                   ++ ")(_inputs_and_cells)"
+                   ++ "\n\n"
+               else ""
+           )
+        -- controller logic
+        ++ controller
+        ++ "\n\n"
+        -- return next cells and outputs (using JS object)
```

Comment:
isnt this just a dict in python?

Reply:
Yea, the input and output to the python function are both dicts.